teps/0094-configuring-resources-at-runtime.md

```diff
@@ -52,12 +61,362 @@ While this may be a valuable feature, it should be considered in a separate prop
-- Users can specify resource requirements for individual `Step`s and `Sidecar`s.
\ No newline at end of file
+- Users can specify resource requirements for individual `Step`s and `Sidecar`s.
+
+## Proposal
+
+Augment `TaskRunSpec` and `PipelineTaskRunSpec` with a mapping of `Step` names to overrides
+and a mapping of `Sidecar` names to overrides.
+
+## Design Details
+
+```go
+import corev1 "k8s.io/api/core/v1"
+
+type TaskRunSpec struct {
+   ...
+   // A map of Step names to ResourceRequirements for that Step.
+   // A single entry with key "*" will apply the ResourceRequirements to all Steps.
+   // If ResourceRequirements are specified in both a Step and StepResources,
+   // the value from the StepResources will be used.
+   StepResources map[string]corev1.ResourceRequirements
+
+   // A map of Sidecar names to ResourceRequirements to apply to that Sidecar.
+   // A single entry with key "*" will apply the ResourceRequirements to all Sidecars.
+   // If ResourceRequirements are specified in both a Sidecar and SidecarResources,
+   // the value from the SidecarResources will be used.
+   SidecarResources map[string]corev1.ResourceRequirements
+}
+
+type PipelineTaskRunSpec struct {
+  ...
+   // A map of Step names to ResourceRequirements for that Step.
+   // A single entry with key "*" will apply the ResourceRequirements to all Steps.
+   // If ResourceRequirements are specified in both a Step and StepResources,
+   // the value from the StepResources will be used.
+   StepResources map[string]corev1.ResourceRequirements
+
+   // A map of Sidecar names to ResourceRequirements to apply to that Sidecar.
+   // A single entry with key "*" will apply the ResourceRequirements to all Sidecars.
+   // If ResourceRequirements are specified in both a Sidecar and SidecarResources,
+   // the value from the SidecarResources will be used.
+   SidecarResources map[string]corev1.ResourceRequirements
+}
+```
+
+### Notes/Caveats
+
+#### Mapping container overrides to Steps and Sidecars
+
+This TEP proposes mapping container overrides to their corresponding `Step`s and `Sidecar`s
+using `Step` and `Sidecar` names. This strategy is consistent with other parts of the Tekton API,
+such as `Pipeline`s and `TaskRun`s referring to `Task`s by their names.
+It also meets the requirement that users can specify resource requirements for individual `Step`s and `Sidecar`s.
+
+Some users may want to have resource requirements apply to every `Step` or `Sidecar`.
+This TEP proposes using "*" as a special string which applies the 
+specified resource requirements to all `Step`s or `Sidecar`s. "*" is not a valid `Container` name.
+
+An alternative option is to use a repeated field for container overrides, where the index
+of the container override is the same as the index of the `Step` or `Sidecar` it applies to.
+This strategy could result in unintended overrides in a `TaskRun`
+if `Step`s are reordered, added, or removed in the corresponding `Task`.
+
```

Comment:
> What happens if the name doesn't match any steps?

I think we should fail?

> What happens if the user's step is unnamed?

I initially though using an index might be an option,  but perhaps not, since the order of steps is not part of the "stable" interface of a task in the catalog, so the index may change with no change in version number.

> What happens if multiple fields match? (e.g. named + *)

I though it was only exact matches?

Reply:
> I though it was only exact matches?

An earlier version proposed using `*` as a "apply to all steps" syntax - this is now an empty string. The same question applies though - e.g. what happens if I do something like:

```yaml
stepOverrides:
        - taskRunStepName: ""
          resources:
            requests:
              memory: 1Gi
        - taskRunStepName: foo
          resources:
            requests:
              memory: 2Gi
        - taskRunStepName: foo
          resources:
            requests:
              cpu: 1
        - taskRunStepName: ""
          resources:
            limits:
              memory: 2Gi
```